src/simd.jl

```diff
@@ -0,0 +1,448 @@
+    return quote
+        $(Expr(:meta, :inline))
+        @inbounds begin
+            D1 = get_data(S1); SV180 = tosimd(D1, Val{1}, Val{80})
+            D2 = get_data(S2); SV280 = tosimd(D2, Val{1}, Val{80})
+            r = SV180 - SV280
+            r81 = D1[81] - D2[81]
+            return Tensor{4, 3}($(Expr(:tuple, [:(r[$i]) for i in 1:80]..., :(r81))))
+        end
+    end
+end
+
+##########################################
+# (2): * and / between tensor and number #
+##########################################
+# note it is allowed with different eltypes, since it is promoted in SIMD.jl
+@generated function Base.:*{T <: SIMDTypes}(n::T, S::AllSIMDTensors{T})
+    TensorType = get_base(S)
+    return quote
+        $(Expr(:meta, :inline))
+        @inbounds begin
+            D = get_data(S); SV = tosimd(D)
+            r = n * SV
+            return $TensorType(r)
+        end
+    end
+end
+@generated function Base.:*{T <: SIMDTypes}(n::T, S::Tensor{4, 3, T})
+    return quote
+        $(Expr(:meta, :inline))
+        @inbounds begin
+            D = get_data(S); SV80 = tosimd(D, Val{1}, Val{80})
+            r = n * SV80; r81 = n * D[81]
+            return Tensor{4, 3}($(Expr(:tuple, [:(r[$i]) for i in 1:80]..., :(r81))))
+        end
+    end
+end
+@generated function Base.:*{T <: SIMDTypes}(S::AllSIMDTensors{T}, n::T)
+    TensorType = get_base(S)
+    return quote
+        $(Expr(:meta, :inline))
+        @inbounds begin
+            D = get_data(S); SV = tosimd(D)
+            r = SV * n
+            return $TensorType(r)
+        end
+    end
+end
+@generated function Base.:*{T <: SIMDTypes}(S::Tensor{4, 3, T}, n::T)
+    return quote
+        $(Expr(:meta, :inline))
+        @inbounds begin
+            D = get_data(S); SV80 = tosimd(D, Val{1}, Val{80})
+            r = SV80 * n; r81 = D[81] * n
+            return Tensor{4, 3}($(Expr(:tuple, [:(r[$i]) for i in 1:80]..., :(r81))))
+        end
+    end
+end
+@generated function Base.:/{T <: SIMDTypes}(S::AllSIMDTensors{T}, n::T)
+    TensorType = get_base(S)
+    return quote
+        $(Expr(:meta, :inline))
+        @inbounds begin
+            D = get_data(S); SV = tosimd(D)
+            r = SV / n
+            return $TensorType(r)
+        end
+    end
+end
+@generated function Base.:/{T <: SIMDTypes}(S::Tensor{4, 3, T}, n::T)
+    return quote
+        $(Expr(:meta, :inline))
+        @inbounds begin
+            D = get_data(S); SV80 = tosimd(D, Val{1}, Val{80})
+            r = SV80 / n; r81 = D[81] / n
+            return Tensor{4, 3}($(Expr(:tuple, [:(r[$i]) for i in 1:80]..., :(r81))))
+        end
+    end
+end
+
+############
+# (3): dot #
+############
+# 2-1
+
+@inline function Base.dot{T <: SIMDTypes, N}(S1::Tensor{2, 2, T, N}, S2::Vec{2, T})
+    @inbounds begin
+        D1 = get_data(S1); D2 = get_data(S2)
+        SV11 = tosimd(D1, Val{1}, Val{2})
+        SV12 = tosimd(D1, Val{3}, Val{4})
+        r = fma(SV12, D2[2], SV11 * D2[1])
```

Comment:
We can just use that then. That's the same for me, and simpler than trying to detect cpu I guess.